### ISSUE TITLE
Upgrade to Parquet filter2 API

### ISSUE THREAD

**Issue by fnothaft** [bigdatagenomics/adam#565]
Currently, we're using the old `UnboundRecordPredicate` API for reading data from Parquet. We should migrate to the newer `filter2` API, which supports block statistics.

**Comment by fnothaft** [comment]
Discussion point: If we did this, could we remove the `org.bdgenomics.adam.predicates` package?

**Comment by tdanford** [comment]
Ping @carlyeks

**Comment by tomwhite** [comment]
The main advantage of the filter2 API is that it filters out entire row groups. This means that queries for a range of positions, say, are very efficient since they only load the relevant row groups. I wrote a test to demonstrate this: https://github.com/bigdatagenomics/adam/compare/master...tomwhite:ADAM-565-filter2. (Note that the test can't assert that only particular row groups were loaded, but I observed this in the debugger.)

On the other hand, there is currently an outstanding Parquet issue to investigate why the filter2 API is slower than the original (https://issues.apache.org/jira/browse/PARQUET-98). It would be good for that to be resolved before switching.

I wasn't sure how to take this further - should the load methods on ADAMContext be changed to take a FilterPredicate? (I.e. is ADAMPredicate being removed?) This would be in line with https://github.com/bigdatagenomics/adam/issues/608.

**Comment by fnothaft** [comment]
Hi @tomwhite! Great to have you on here. The new statistics functionality in filter2 is pretty excellent. I'm hopeful that if we moved over to filter2, we'd be able to avoid needing indices to serve very small slices of data (e.g., everyone's favorite "look up dbSNP site RS12314" queries).

> I wasn't sure how to take this further - should the load methods on ADAMContext be changed to take a FilterPredicate? (I.e. is ADAMPredicate being removed?) This would be in line with #608.

I personally would like to remove `ADAMPredicate`. If we did go whole hog with #608, `ADAMPredicate` could be removed trivially. I'm about halfway in agreement with #608. I would prefer to maintain the predicate functions on the "magic" functions, but I suppose I can be swayed. That being said, it's not clear if we _could_ remove `ADAMPredicate` without going all-in on #608.

In short,  I may be starting to agree with @laserson on #608.

> On the other hand, there is currently an outstanding Parquet issue to investigate why the filter2 API is slower than the original (https://issues.apache.org/jira/browse/PARQUET-98). It would be good for that to be resolved before switching.

IMO, the speedup for `filter` is fairly tame for us (I've measured 1.3x speedup when filtering 50% of data, 1.8x speedup when filtering 99%). I think it'd be vastly preferable to be able to drop row groups at the cost of predicate performance when filtering in a row group. Might make sense to move towards `filter2` and then benchmark? From your test code, it looks like it wouldn't be too huge of a change.

**Comment by nikhilRP** [comment]
+1

**Comment by tomwhite** [comment]
> I'm hopeful that if we moved over to filter2, we'd be able to avoid needing indices to serve very small slices of data (e.g., everyone's favorite "look up dbSNP site RS12314" queries).

Exactly.

> If we did go whole hog with #608, ADAMPredicate could be removed trivially.

I'll have a go at replacing ADAMPredicate with FilterPredicate in ADAMContext  and see what that looks like.

**Comment by tomwhite** [comment]
I've opened a PR at https://github.com/bigdatagenomics/adam/pull/616. Please let me know what you think.

**Comment by fnothaft** [comment]
Closed by #616.
